server/org.eclipse.emfcloud.ecore.glsp/src/main/java/org/eclipse/emfcloud/ecore/glsp/EcoreModelIndex.java

```diff
@@ -58,6 +62,39 @@ public void indexSemantic(String id, EObject semanticElement) {
 		semanticIndex.putIfAbsent(id, semanticElement);
 	}
 
+	public void indexGModelElement(String uri, GModelElement gModelElement){
+		uriToGModelElement.putIfAbsent(uri, gModelElement);
+	}
+
+	public void removeURI(String uri){
+		uriToGModelElement.remove(uri);
+	}
+
+	public void updateURI(String uri, String newUri){
+		GModelElement element = uriToGModelElement.get(uri);
+		removeURI(uri);
+		indexGModelElement(newUri, element);
+	}
+
+	public void updateReferenceRoot(String uri, String newUri){
+		for(String s: uriToGModelElement.keySet()){
+			if(s.substring(2, 2 + uri.length()).equals(uri) && s.charAt(uri.length() + 2) == '/'){
```

Comment:
Too much magic replacements here; you should 1) move this to a separate method and 2) document the expected format for URIs. This will probably fail for any URI that doesn't have the expected format.
Using the EMF URI class (URI.createURI(string) and uri.toFileString()) might be cleaner and more reliable, depending on which kind of URI is expected